test/test_code_generation/test_models_composition.py

```diff
@@ -21,71 +21,70 @@ def test_list_ex():
+        "id": input_dict["id"] + "_list",
+        "value": []
+    }
+
+    for item in input_dict["value"]:
+        # item is a tuple of model name and model data
+        model = [{key: deepcopy(value)} for key, value in item[1].items()]
+        outputs["value"].append((item[0], model))
+
+    return outputs
+
+
 # This test relies on model names as a some sort of models ids
 # and may fail if some logic of their generation will be changed
 # List of Tuple[root_model_name, JSON data] | Dict[model_name, Set[root_model_names]]
-test_extract_root_data = [
-    pytest.param(
-        [
-            ("TestModelA", {
-                "count": 1,
-                "items": [
-                    {
-                        "x": .5,
-                        "y": .1
-                    }
-                ]
-            }),
-            ("TestModelB", {
-                "next": "some_url",
-                "prev": None,
-                "count": 2000,
-                "items": [
-                    {
-                        "x": .5,
-                        "y": .1
-                    }
-                ]
-            }),
+extract_root_data_input = [
+    {
+        "value": [
+            ("TestModelA", {"count": 1, "items": [{"x": 0.5, "y": 0.1}]}),
+            (
+                "TestModelB",
+                {
+                    "next": "some_url",
+                    "prev": None,
+                    "count": 2000,
+                    "items": [{"x": 0.5, "y": 0.1}],
+                },
+            ),
         ],
-        {
-            'Item': {'TestModelA', 'TestModelB'},
-            'TestModelA': set()
-        }
-    ),
-    pytest.param(
-        [
-            ("TestModelA", {
-                "count": 1,
-                "items": [
-                    {
-                        "x": .5,
-                        "y": .1
-                    }
-                ]
-            }),
-            ("TestModelB", {
-                "count": 1,
-                "items": [
-                    {
-                        "x": .5,
-                        "y": .1
-                    }
-                ]
-            }),
+        "expected": {"Item": {"TestModelA", "TestModelB"}, "TestModelA": set()},
+        "id": "separate_roots"
+    },{
+        "value": [
+            ("TestModelA", {"count": 1, "items": [{"x": 0.5, "y": 0.1}]}),
+            ("TestModelB", {"count": 1, "items": [{"x": 0.5, "y": 0.1}]}),
         ],
-        {
-            'Item': {'TestModelA_TestModelB'},
-            'TestModelA_TestModelB': set()
-        },
-        id="merge_root"
-    )
-]
+        "expected": {"Item": {"TestModelA_TestModelB"}, "TestModelA_TestModelB": set()},
+        "id": "merge_root",
+    }]
 
+extract_root_data_input_list = [generate_list_input(i) for i in extract_root_data_input]
+test_extract_root_data = [pytest.param(inpt["value"], inpt["expected"], id=inpt["id"]) for inpt in extract_root_data_input + extract_root_data_input_list]
```

Comment:
Small reorganisation of the test data to test both dict and list input. Removing the fix will make the tests with  data from `extract_root_data_input_list` fail.